theano/gpuarray/type.py

```diff
@@ -250,6 +260,12 @@ def filter(self, data, strict=False, allow_downcast=None):
             if b and shp[i] != 1:
                 raise TypeError("Non-unit value on shape on a broadcastable"
                                 " dimension.", shp, self.broadcastable)
+        if not isinstance(data, gpuarray.GpuArray):
+            if old_data is not None and old_data.shape == data.shape:
+                old_data.write(data)
+                data = old_data
+            else:
+                data = pygpu.array(data, context=self.context)
         return data
```

Comment:
The old back-end currently have the same problem. If we call set_value(a_gpu_array, borrow=False), then in GpuArraySharedVariable.set_value(), we will do a copy before calling filter_inplace. And filter_inplace won't copy it to the old memory region. So we end up doing an extra allocation for no good reason.

There isn't an easy fix. It will probably need to add a method set_value to the class Container in gof/link.py. I'll make an issue about that. This isn't for this PR.